application/ui/src/shared/annotator/annotation-actions-provider.component.tsx

```diff
@@ -55,104 +56,78 @@ const AnnotationsContext = createContext<AnnotationsContextValue | null>(null);
 
 type AnnotationActionsProviderProps = {
     children: ReactNode;
+    initialAnnotationsDTO?: ServerAnnotation[];
+    isUserReviewed?: boolean;
     mediaItem: DatasetItem;
 };
 
-const isUnannotatedError = (error: unknown): boolean => {
-    return (
-        isObject(error) && 'detail' in error && /Dataset item has not been annotated yet/i.test(String(error.detail))
-    );
-};
-
-export const AnnotationActionsProvider = ({ children, mediaItem }: AnnotationActionsProviderProps) => {
+export const AnnotationActionsProvider = ({
+    children,
+    initialAnnotationsDTO = [],
+    isUserReviewed = false,
+    mediaItem,
+}: AnnotationActionsProviderProps) => {
     const projectId = useProjectIdentifier();
     const saveMutation = $api.useMutation(
         'post',
         '/api/projects/{project_id}/dataset/items/{dataset_item_id}/annotations'
     );
 
-    const { data: serverAnnotations, error: fetchError } = $api.useQuery(
-        'get',
-        '/api/projects/{project_id}/dataset/items/{dataset_item_id}/annotations',
-        {
-            params: { path: { project_id: projectId, dataset_item_id: mediaItem.id || '' } },
-        },
-        {
-            retry: (_failureCount, error: unknown) => !isUnannotatedError(error),
-        }
-    );
-
     const { data: project } = useProject();
 
-    const [localAnnotations, setLocalAnnotations] = useState<Annotation[]>([]);
-    const isDirty = useRef<boolean>(false);
+    const localFormattedAnnotations = useMemo(() => {
+        const projectLabels = project?.task?.labels || [];
+
+        return mapServerAnnotationsToLocal(initialAnnotationsDTO, projectLabels);
+    }, [initialAnnotationsDTO, project?.task?.labels]);
+
+    const [annotations, setAnnotations, undoRedoActions] = useUndoRedoState<Annotation[]>([]);
+
+    useEffect(() => {
+        if (localFormattedAnnotations.length > 0) {
+            undoRedoActions.reset(localFormattedAnnotations);
+        }
+    }, [localFormattedAnnotations, undoRedoActions]);
```

Comment:
Do we really need this use effect? I.e. wouldn't the state be reset automatically due to the `key={mediaItem.id}` that is set on the `AnnotationActionsProvider`?

Alternatively I'd move the memo stuff inside of the useEffect. Another thing that might help is to extract this into a dedicated hook (I found `localFormattedAnnotatiosn` to be a rather weirdly named variable which made me confused :D )

Maybe something like this,

```typescript
const useResetAnnotations(initialAnnotations: AnnotationsDTO, projectLabels, undoRedoActions) = () => {
    useEffect(() => {
        const annotations = mapServerAnnotationsToLocal(initialAnnotations, projectLabels ?? []);

        if (annotations.length > 0) {
            undoRedoActions.reset(annotations);
        }
    }, [initialAnnotationsDTO, project?.task?.labels, undoRedoActions]);
}
```

and use like

```suggestion
    const [annotations, setAnnotations, undoRedoActions] = useUndoRedoState<Annotation[]>([]);
    useResetAnnotations(initialAnnotationsDTO, project?.task?.labels, undoRedoActions)
```